synapse/handlers/sync.py

```diff
@@ -1314,36 +1314,89 @@ async def generate_sync_result(
         At the end, we transfer data from the `sync_result_builder` to a new `SyncResult`
         instance to signify that the sync calculation is complete.
         """
+
+        user_id = sync_config.user.to_string()
+        app_service = self.store.get_app_service_by_user_id(user_id)
+        if app_service:
+            # We no longer support AS users using /sync directly.
+            # See https://github.com/matrix-org/matrix-doc/issues/1144
+            raise NotImplementedError()
+
+        # Note: we get the users room list *before* we get the current token, this
+        # avoids checking back in history if rooms are joined after the token is fetched.
+        mutable_joined_room_ids = set(await self.store.get_rooms_for_user(user_id))
+
         # NB: The now_token gets changed by some of the generate_sync_* methods,
         # this is due to some of the underlying streams not supporting the ability
         # to query up to a given point.
         # Always use the `now_token` in `SyncResultBuilder`
         now_token = self.event_sources.get_current_token()
         log_kv({"now_token": now_token})
 
+        # Since we fetched the users room list before the token, there's a small window
+        # during which membership events may have been persisted, so we fetch these now
+        # and modify the joined room list for any changes between the get_rooms_for_user
+        # call and the get_current_token call.
+        membership_change_events = []
+        if since_token:
+            membership_change_events = await self.store.get_membership_changes_for_user(
+                user_id, since_token.room_key, now_token.room_key, self.rooms_to_exclude
+            )
+
+            mem_last_change_by_room_id: Dict[str, EventBase] = {}
+            for event in membership_change_events:
+                mem_last_change_by_room_id[event.room_id] = event
+
+            # For the latest membership event in each room found, add/remove the room ID
+            # from the joined room list accordingly. In this case we only care if the
+            # latest change is JOIN.
+
+            for room_id, event in mem_last_change_by_room_id.items():
+                logger.info(
+                    "User membership change between getting rooms and current token: %s %s %s",
+                    user_id,
+                    event.membership,
+                    room_id,
+                )
+                # User joined a room - we have to then check the room state to ensure we
+                # respect any bans if there's a race between the join and ban events.
+                if event.membership == Membership.JOIN:
```

Comment:
Copied from old draft PR

**[erikjohnston](https://github.com/erikjohnston) [3 days ago](https://github.com/beeper/synapse/pull/11/files#r980869272)**

I think we need to be careful that we handle a join -> join transition? e.g. a display name update?

I think we could look at the current_state_deltas table to get a more accurate sense of how the state has changed?

**[erikjohnston](https://github.com/erikjohnston) erikjohnston [3 days ago](https://github.com/beeper/synapse/pull/11/files#r980870913)**

Actually, no we don't need to worry about join -> join transitions. I was thinking of a different un-applicable scenario.

current_state_deltas probably still the right table?
Member Author

**[Fizzadar](https://github.com/Fizzadar) Fizzadar [2 days ago](https://github.com/beeper/synapse/pull/11/files#r982135756)**

Was just trying to figure this out but I don't think current_state_deltas has all the information we need - specifically the membership field in event content means we can't determine whether an event was a join or invite.

Pulling in the events would solve this, but at that point it's essentially the same as the current get_membership_changes_for_user call I think, is there a reason this isn't suitable?

Reply:
`current_state_deltas` will give you the actual changes in current state between two tokens, whereas `get_membership_changes_for_user` is a bit dodge and only gets an membership events that have been persisted in the range (these are technically not the same thing).

> Pulling in the events would solve this, but at that point it's essentially the same as the current `get_membership_changes_for_user` call I think

You'd want to join against `room_memberships`, so `current_state_deltas` may be cheaper as there will likely be less state changes than events in the room.

Though it occurs to me that we only keep `current_state_deltas` for a month, so we'd have to handle the case that the since token given was too old. Ideally tell the client to do a full resync, but at that point we're creeping the scope of this PR a bunch

Reply:
Is there a way to ask clients to re-init sync currently? Assuming this is possible I think all the membership change lookups in sync could be optimised out to use the `current_stream_delta` x `room_memberships` and drop the `get_membership_changes_for_user` calls entirely. I wonder/hope if this could dramatically simplify the [current membership change handling](https://github.com/matrix-org/synapse/blob/develop/synapse/handlers/sync.py#L1897-L2042) in sync.

I think this fits into a subsequent PR though, definitely out of scope here! As it stands I believe this change will behave as expected; the sync handler already uses the membership change events to calculate room changes.

Reply:
Yeah, sliding sync supports it